File: ProblemSet/Week 3/Problem/JavaScript/LinkedList/src/auction.js
class Auction {
    constructor(item, startPrice, duration, description) {
        this.item = item;
        this.startPrice = startPrice;
        this.duration = duration * 60000; // Convert minutes to milliseconds
        this.description = description;
        this.bids = [];
        this.startTime = Date.now();
    }

    placeBid(bidder, amount) {
        if (amount > this.getCurrentPrice() && this.isActive()) {
            this.bids.push({ bidder, amount });
            return true;
        }
        return false;
    }

    isActive() {
        const elapsedTime = Date.now() - this.startTime;
        return elapsedTime < this.duration;
    }

    getCurrentPrice() {
        return this.bids.length > 0 ? this.bids[this.bids.length - 1].amount : this.startPrice;
    }

    getDescription() {
        return this.description;
    }

    getWinner() {
        if (this.bids.length > 0) {
            return this.bids.reduce((highestBid, currentBid) => {
                return currentBid.amount > highestBid.amount ? currentBid : highestBid;
            });
        }
        return null;
    }
}

module.exports = Auction;

File: ProblemSet/Week 3/Problem/Python/LinkedList-v2-OOP/blockchain_auction_system/auction.py
from blockchain_auction_system.double_ended_linked_list import DoubleEndedLinkedList


class Auction:
    """
    Represents an auction in the system.
    """
    def __init__(self, item, start_price, end_time):
        self.item = item
        self.start_price = start_price
        self.current_price = start_price
        self.end_time = end_time
        self.bids = DoubleEndedLinkedList()

    def place_bid(self, bidder, amount):
        """
        Place a bid in the auction.
        """
        if amount > self.current_price and time.time() < self.end_time:
            self.current_price = amount
            self.bids.insert_at_end((bidder, amount))
            return True
        return False